{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false
   },
   "source": [
    "## SampleSheet.py"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false
   },
   "source": [
    "https://github.com/YamamotoLabUCSF/SampleSheet  \n",
    "v1.0/Committed 08-01-2019\n",
    "\n",
    "<img src=\"SampleSheet_img/SampleSheet_thumbnail.png\" align=\"left\" width=\"600\">\n",
    "\n",
    "### Background:\n",
    "Sequencing by synthesis (SBS) collects millions to billions of DNA sequence reads *en masse*. DNA templates from tens to thousands of independent sample sources can be barcoded, pooled, and sequenced on a common flow cell. Unique indices (barcodes) allow pooled reads to be re-assigned to cognate sample sources (demultiplexed).\n",
    "\n",
    "**This script automates creation of an Illumina® Sample Sheet, the comma-separated text document required by Illumina® sequencing systems to specify (1) sequencing parameters and (2) sample-barcode relationships.** With this script, a Sample Sheet with up to 9,216 sample-barcode relationships can be automatically generated in <1 second, following user entry of a single list containing up to 96 sample prefixes assigned to an i7 index range and unique i5 index (each sample prefix to be expanded to up to 96 individual samples, suffixed by well ID (*i.e.*, A01-H12) of a 96-well plate).\n",
    "\n",
    "### Synopsis:\n",
    "**This script returns a Sample Sheet file compatible with Illumina® sequencing platforms.**  \n",
    "\n",
    "**Its barcode assignments depend on a set of 192 primers with unique 8-bp barcodes compatible with Illumina® sequencing platforms; these barcodes allow up to 9,216 samples to be arrayed in 96-well (or 384-well) format with unique barcodes for pooled sequencing.**\n",
    "\n",
    "**Users are asked for the path to an output directory in which a Sample Sheet will be created, along with user-specific variables for Sample Sheet \\[Header\\], \\[Reads\\], and \\[Data\\] sections.**\n",
    ">(see 'Input notes' for details).\n",
    "    \n",
    "Python3 is required for operation.  \n",
    "\n",
    "Note on index usage: In this script, i7 is designated for use in full plate format (each well is uniquely barcoded by a single i7 index), whereas i5 defines all wells of a specific plate (up to 96 wells in a single plate are barcoded by a common i5 index).\n",
    "\n",
    "For usage details, please refer to README file at GitHub and to the following manuscript:  \n",
    ">*Ehmsen, Knuesel, Martinez, Asahina, Aridomi, Yamamoto (2020)*\n",
    "    \n",
    "Please cite usage as:  \n",
    ">ImputedGenotypes.py  \n",
    ">*Ehmsen, Knuesel, Martinez, Asahina, Aridomi, Yamamoto (2020)*\n",
    " \n",
    "--------\n",
    "\n",
    "### Operation notes:\n",
    "*What does this script do?*\n",
    "\n",
    "This script automates creation of a Sample Sheet compatible with Illumina® sequencing, based on custom i7 (96) and i5 (96) index sequences used to barcode up to 9,216 distinct samples. Specifically, the script performs these operations:\n",
    "\n",
    " 1. **collect user input**\n",
    "    - specify Dual Indexed Sequencing Workflow (A *vs.* B)\n",
    "    - specify Investigator Name, Project Name\n",
    "    - specify single index (SE) *vs.* dual-indexed (PE) barcode format\n",
    "    - collect Data relationships (sample ID, i7 barcode range (1-96) and single i5 barcode (1-96)  \n",
    " \n",
    " \n",
    " 2. **generate Sample Sheet**\n",
    "    - populate \\[Header\\] and \\[Reads\\] sections based on user input for Investigator Name, Project Name, SE/PE format\n",
    "    - populate \\[Data\\] section based on expansion of i7 barcode range and i5 barcode designation for each 96-well plate  \n",
    "    (appropriate i7 and i5 index sequences are populated based on user-specified Workflow (A *vs.* B))\n",
    "\n",
    "### Input notes:\n",
    "You will be prompted for the following user-specific information:\n",
    "\n",
    "**Required** (4 strings): \n",
    "      <ul>\n",
    "      <li>where should output file go?</li>\n",
    "          *absolute path to **output directory** and filename for Sample Sheet*\n",
    "      <li>Investigator Name</li>\n",
    "          *character string specifying name to be associated with Sample Sheet and sequencing run*         \n",
    "      <li>Project Name</li>\n",
    "          *character string specifying project name to be associated with Sample Sheet and sequencing run*\n",
    "      <li>Run type specification: Single-end (SE) or Paired-end (PE) sequencing run?  How many sequencing cycles (read length for R1 (Read 1) and R2 (Read 2))?\n",
    "          *...comma-separated character string indicating SE vs. PE, # of sequencing cycles (R1), # of sequencing cycles (R2, if applicable)*\n",
    "      <li>List of sample:barcode relationships</li>\n",
    "          *single lines of comma-separated character strings specifying overarching sample prefix to assign to up to 96 samples arrayed in 96-well plate format (prefix is parsed to samples with well suffixes, e.g., -A01, -A02...-H12); barcode assignments (i7 and i5) are designated to individual samples based on integer range (i7) or integer (i5) assigned to plate*\n",
    "      </ul>\n",
    "  \n",
    "Note on list of sample:barcode relationships: This is a list of plate names (prefixes), i7 index range, and i5 index.  \n",
    "For example: 'DG-1, 1-96, 5' on a single line of text would indicate plate name/prefix 'DG-1' applied to up to 96 samples (uniqued identified by well position A01-H12, *e.g.*, DG-1-A01, DG-1-A02, ... DG-1-H12), range of i7 indices used to barcode individual wells in this 96-well plate (*e.g.*, A01-H12), and i5 index used across all wells of this plate (*e.g.*, A05).\n",
    "\n",
    "### Output notes:\n",
    "In brief: Illumina® Sample Sheets accommodate up to 10 column fields, but only 5 of these (fields 2, 5-8) are required for a sequencing run (indicated below).  This script outputs only these 5 required column fields.  \n",
    " \n",
    "10 Sample Sheet column fields:  \n",
    "  - field 1: Sample_ID,  \n",
    "  - field 2: Sample_Name,  \n",
    "  - field 3: Sample_Plate,  \n",
    "  - field 4: Sample_Well,  \n",
    "  - field 5: I7_Index_ID, \n",
    "  - field 6: index,  \n",
    "  - field 7: I5_Index_ID,  \n",
    "  - field 8: index2, \n",
    "  - field 9: Sample_Project,  \n",
    "  - field 10: Description  \n",
    "\n",
    "Fields customized and output by this script:  \n",
    "  - field 2: Sample_Name,  \n",
    "  - field 5: I7_Index_ID, \n",
    "  - field 6: I7 index sequence,  \n",
    "  - field 7: I5_Index_ID,  \n",
    "  - field 8: I5 index sequence  \n",
    " \n",
    "--------\n",
    "### Visual summary of key script operations:\n",
    "In short, **brief user inputs** (*e.g.,* below), are converted to **Sample Sheet** contents compatible with Illumina® sequencing (**key output file**, below). In particular, a minimal list of up to 96 \\[Data\\] relationships is expanded in microseconds to a \\[Data\\] section containing up to 9,216 sample:barcode relationships.  \n",
    "  \n",
    "*example*  \n",
    "\n",
    "------\n",
    "**input:**  \n",
    "*Path to Sample Sheet file name for creation:* /Users/name/SampleSheet.csv  \n",
    "*Workflow:* A  \n",
    "*Investigator Name, Project Name*: Dorothy Gale, Sequencing  \n",
    "*SE vs. PE, cycle details*: PE, 151, 151  \n",
    "*sample:barcoded relationships:*  \n",
    "DG-1, 1-96, 4  \n",
    "DG-2, 1-96, 9  \n",
    "DG-2, 1-96, 78  \n",
    "<br clear=\"all\" />\n",
    "**output:**\n",
    "<br clear=\"all\" />\n",
    "<img src=\"SampleSheet_img/SampleSheet_example.png\" align=\"mid\" width=\"500\">  \n",
    "<div style=text-align:center>...etc.</div>\n",
    "\n",
    "-------------------------------------------------------------------------------\n",
    "**Welcome.**  \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false
   },
   "source": [
    "### I. Setup  \n",
    "Import libraries, modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deletable": false,
    "editable": false,
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Operating system interfaces\n",
    "import os\n",
    "\n",
    "# Object-oriented filesystem paths\n",
    "from pathlib import Path\n",
    "\n",
    "# System-specific parameters and functions\n",
    "import sys\n",
    "\n",
    "# Implementation of import\n",
    "import importlib\n",
    "\n",
    "# Time access and conversions, Basic data and time types\n",
    "import time\n",
    "from datetime import datetime\n",
    "\n",
    "# IPython module: display\n",
    "from IPython.display import Image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deletable": false,
    "editable": false,
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Log start time\n",
    "initialTime = datetime.now()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false
   },
   "source": [
    "Define lists and dictionaries that establish integer (1-96) and wellID (A01-H12), and wellID and barcode relationships."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deletable": false,
    "editable": false,
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Define 3 lists and 4 dictionaries that this script will use to establish relationships between well_ID number (1-96) and well_ID label (A01-H12) [well_ID_list], between well_ID number (1-96) and index name (i5A01-i5H12 or i7A01-i7H12) [i5_well_IDs and i7_well_IDs], and between index name (i5A01-i5H12 or i7A01-i7H12) and index sequence (8-bp unique sequence) [i5Dict and i7Dict, i5_revcomp_Dict, i7_revcomp_Dict]:\n",
    "\n",
    "well_ID_list = [\n",
    "('1','A01'),\n",
    "('2','A02'),\n",
    "('3','A03'),\n",
    "('4','A04'),\n",
    "('5','A05'),\n",
    "('6','A06'),\n",
    "('7','A07'),\n",
    "('8','A08'),\n",
    "('9','A09'),\n",
    "('10','A10'),\n",
    "('11','A11'),\n",
    "('12','A12'),\n",
    "('13','B01'),\n",
    "('14','B02'),\n",
    "('15','B03'),\n",
    "('16','B04'),\n",
    "('17','B05'),\n",
    "('18','B06'),\n",
    "('19','B07'),\n",
    "('20','B08'),\n",
    "('21','B09'),\n",
    "('22','B10'),\n",
    "('23','B11'),\n",
    "('24','B12'),\n",
    "('25','C01'),\n",
    "('26','C02'),\n",
    "('27','C03'),\n",
    "('28','C04'),\n",
    "('29','C05'),\n",
    "('30','C06'),\n",
    "('31','C07'),\n",
    "('32','C08'),\n",
    "('33','C09'),\n",
    "('34','C10'),\n",
    "('35','C11'),\n",
    "('36','C12'),\n",
    "('37','D01'),\n",
    "('38','D02'),\n",
    "('39','D03'),\n",
    "('40','D04'),\n",
    "('41','D05'),\n",
    "('42','D06'),\n",
    "('43','D07'),\n",
    "('44','D08'),\n",
    "('45','D09'),\n",
    "('46','D10'),\n",
    "('47','D11'),\n",
    "('48','D12'),\n",
    "('49','E01'),\n",
    "('50','E02'),\n",
    "('51','E03'),\n",
    "('52','E04'),\n",
    "('53','E05'),\n",
    "('54','E06'),\n",
    "('55','E07'),\n",
    "('56','E08'),\n",
    "('57','E09'),\n",
    "('58','E10'),\n",
    "('59','E11'),\n",
    "('60','E12'),\n",
    "('61','F01'),\n",
    "('62','F02'),\n",
    "('63','F03'),\n",
    "('64','F04'),\n",
    "('65','F05'),\n",
    "('66','F06'),\n",
    "('67','F07'),\n",
    "('68','F08'),\n",
    "('69','F09'),\n",
    "('70','F10'),\n",
    "('71','F11'),\n",
    "('72','F12'),\n",
    "('73','G01'),\n",
    "('74','G02'),\n",
    "('75','G03'),\n",
    "('76','G04'),\n",
    "('77','G05'),\n",
    "('78','G06'),\n",
    "('79','G07'),\n",
    "('80','G08'),\n",
    "('81','G09'),\n",
    "('82','G10'),\n",
    "('83','G11'),\n",
    "('84','G12'),\n",
    "('85','H01'),\n",
    "('86','H02'),\n",
    "('87','H03'),\n",
    "('88','H04'),\n",
    "('89','H05'),\n",
    "('90','H06'),\n",
    "('91','H07'),\n",
    "('92','H08'),\n",
    "('93','H09'),\n",
    "('94','H10'),\n",
    "('95','H11'),\n",
    "('96','H12')\n",
    "]\n",
    "\n",
    "i5_well_IDs = [\n",
    "('1','i5A01'),\n",
    "('2','i5A02'),\n",
    "('3','i5A03'),\n",
    "('4','i5A04'),\n",
    "('5','i5A05'),\n",
    "('6','i5A06'),\n",
    "('7','i5A07'),\n",
    "('8','i5A08'),\n",
    "('9','i5A09'),\n",
    "('10','i5A10'),\n",
    "('11','i5A11'),\n",
    "('12','i5A12'),\n",
    "('13','i5B01'),\n",
    "('14','i5B02'),\n",
    "('15','i5B03'),\n",
    "('16','i5B04'),\n",
    "('17','i5B05'),\n",
    "('18','i5B06'),\n",
    "('19','i5B07'),\n",
    "('20','i5B08'),\n",
    "('21','i5B09'),\n",
    "('22','i5B10'),\n",
    "('23','i5B11'),\n",
    "('24','i5B12'),\n",
    "('25','i5C01'),\n",
    "('26','i5C02'),\n",
    "('27','i5C03'),\n",
    "('28','i5C04'),\n",
    "('29','i5C05'),\n",
    "('30','i5C06'),\n",
    "('31','i5C07'),\n",
    "('32','i5C08'),\n",
    "('33','i5C09'),\n",
    "('34','i5C10'),\n",
    "('35','i5C11'),\n",
    "('36','i5C12'),\n",
    "('37','i5D01'),\n",
    "('38','i5D02'),\n",
    "('39','i5D03'),\n",
    "('40','i5D04'),\n",
    "('41','i5D05'),\n",
    "('42','i5D06'),\n",
    "('43','i5D07'),\n",
    "('44','i5D08'),\n",
    "('45','i5D09'),\n",
    "('46','i5D10'),\n",
    "('47','i5D11'),\n",
    "('48','i5D12'),\n",
    "('49','i5E01'),\n",
    "('50','i5E02'),\n",
    "('51','i5E03'),\n",
    "('52','i5E04'),\n",
    "('53','i5E05'),\n",
    "('54','i5E06'),\n",
    "('55','i5E07'),\n",
    "('56','i5E08'),\n",
    "('57','i5E09'),\n",
    "('58','i5E10'),\n",
    "('59','i5E11'),\n",
    "('60','i5E12'),\n",
    "('61','i5F01'),\n",
    "('62','i5F02'),\n",
    "('63','i5F03'),\n",
    "('64','i5F04'),\n",
    "('65','i5F05'),\n",
    "('66','i5F06'),\n",
    "('67','i5F07'),\n",
    "('68','i5F08'),\n",
    "('69','i5F09'),\n",
    "('70','i5F10'),\n",
    "('71','i5F11'),\n",
    "('72','i5F12'),\n",
    "('73','i5G01'),\n",
    "('74','i5G02'),\n",
    "('75','i5G03'),\n",
    "('76','i5G04'),\n",
    "('77','i5G05'),\n",
    "('78','i5G06'),\n",
    "('79','i5G07'),\n",
    "('80','i5G08'),\n",
    "('81','i5G09'),\n",
    "('82','i5G10'),\n",
    "('83','i5G11'),\n",
    "('84','i5G12'),\n",
    "('85','i5H01'),\n",
    "('86','i5H02'),\n",
    "('87','i5H03'),\n",
    "('88','i5H04'),\n",
    "('89','i5H05'),\n",
    "('90','i5H06'),\n",
    "('91','i5H07'),\n",
    "('92','i5H08'),\n",
    "('93','i5H09'),\n",
    "('94','i5H10'),\n",
    "('95','i5H11'),\n",
    "('96','i5H12')\n",
    "]\n",
    "\n",
    "i7_well_IDs = [\n",
    "('1','i7A01'),\n",
    "('2','i7A02'),\n",
    "('3','i7A03'),\n",
    "('4','i7A04'),\n",
    "('5','i7A05'),\n",
    "('6','i7A06'),\n",
    "('7','i7A07'),\n",
    "('8','i7A08'),\n",
    "('9','i7A09'),\n",
    "('10','i7A10'),\n",
    "('11','i7A11'),\n",
    "('12','i7A12'),\n",
    "('13','i7B01'),\n",
    "('14','i7B02'),\n",
    "('15','i7B03'),\n",
    "('16','i7B04'),\n",
    "('17','i7B05'),\n",
    "('18','i7B06'),\n",
    "('19','i7B07'),\n",
    "('20','i7B08'),\n",
    "('21','i7B09'),\n",
    "('22','i7B10'),\n",
    "('23','i7B11'),\n",
    "('24','i7B12'),\n",
    "('25','i7C01'),\n",
    "('26','i7C02'),\n",
    "('27','i7C03'),\n",
    "('28','i7C04'),\n",
    "('29','i7C05'),\n",
    "('30','i7C06'),\n",
    "('31','i7C07'),\n",
    "('32','i7C08'),\n",
    "('33','i7C09'),\n",
    "('34','i7C10'),\n",
    "('35','i7C11'),\n",
    "('36','i7C12'),\n",
    "('37','i7D01'),\n",
    "('38','i7D02'),\n",
    "('39','i7D03'),\n",
    "('40','i7D04'),\n",
    "('41','i7D05'),\n",
    "('42','i7D06'),\n",
    "('43','i7D07'),\n",
    "('44','i7D08'),\n",
    "('45','i7D09'),\n",
    "('46','i7D10'),\n",
    "('47','i7D11'),\n",
    "('48','i7D12'),\n",
    "('49','i7E01'),\n",
    "('50','i7E02'),\n",
    "('51','i7E03'),\n",
    "('52','i7E04'),\n",
    "('53','i7E05'),\n",
    "('54','i7E06'),\n",
    "('55','i7E07'),\n",
    "('56','i7E08'),\n",
    "('57','i7E09'),\n",
    "('58','i7E10'),\n",
    "('59','i7E11'),\n",
    "('60','i7E12'),\n",
    "('61','i7F01'),\n",
    "('62','i7F02'),\n",
    "('63','i7F03'),\n",
    "('64','i7F04'),\n",
    "('65','i7F05'),\n",
    "('66','i7F06'),\n",
    "('67','i7F07'),\n",
    "('68','i7F08'),\n",
    "('69','i7F09'),\n",
    "('70','i7F10'),\n",
    "('71','i7F11'),\n",
    "('72','i7F12'),\n",
    "('73','i7G01'),\n",
    "('74','i7G02'),\n",
    "('75','i7G03'),\n",
    "('76','i7G04'),\n",
    "('77','i7G05'),\n",
    "('78','i7G06'),\n",
    "('79','i7G07'),\n",
    "('80','i7G08'),\n",
    "('81','i7G09'),\n",
    "('82','i7G10'),\n",
    "('83','i7G11'),\n",
    "('84','i7G12'),\n",
    "('85','i7H01'),\n",
    "('86','i7H02'),\n",
    "('87','i7H03'),\n",
    "('88','i7H04'),\n",
    "('89','i7H05'),\n",
    "('90','i7H06'),\n",
    "('91','i7H07'),\n",
    "('92','i7H08'),\n",
    "('93','i7H09'),\n",
    "('94','i7H10'),\n",
    "('95','i7H11'),\n",
    "('96','i7H12')\n",
    "]\n",
    "\n",
    "i5revcomp_Dict = {\n",
    "'i5A01':'CTCCATCA',\n",
    "'i5A02':'CGAATTGA',\n",
    "'i5A03':'CGTTAAGA',\n",
    "'i5A04':'AGGAGTGA',\n",
    "'i5A05':'TCAATCGA',\n",
    "'i5A06':'AGTACCGA',\n",
    "'i5A07':'TCGGATTA',\n",
    "'i5A08':'TTGTGAGA',\n",
    "'i5A09':'TCTCGAGA',\n",
    "'i5A10':'CCGGTATA',\n",
    "'i5A11':'ATCTTGCA',\n",
    "'i5A12':'CCAACTTA',\n",
    "'i5B01':'CTAGGATA',\n",
    "'i5B02':'ACCAGAGA',\n",
    "'i5B03':'AGGTCAGA',\n",
    "'i5B04':'ATCGCATA',\n",
    "'i5B05':'CGCTAGTA',\n",
    "'i5B06':'TCGATGTA',\n",
    "'i5B07':'TGAGTAGA',\n",
    "'i5B08':'AACGCTCA',\n",
    "'i5B09':'TTCGCTGA',\n",
    "'i5B10':'ACGTGCTA',\n",
    "'i5B11':'ATACGCCA',\n",
    "'i5B12':'AGCAGCTA',\n",
    "'i5C01':'TGTGAACA',\n",
    "'i5C02':'ATGCCGTA',\n",
    "'i5C03':'AGAGCACA',\n",
    "'i5C04':'CGTCTACA',\n",
    "'i5C05':'AGGCAACA',\n",
    "'i5C06':'AAGCGTCA',\n",
    "'i5C07':'CCTTATCA',\n",
    "'i5C08':'CGTTGTTA',\n",
    "'i5C09':'TGACGCTA',\n",
    "'i5C10':'ACTCATGA',\n",
    "'i5C11':'TGTGCTTA',\n",
    "'i5C12':'CACCGTTA',\n",
    "'i5D01':'CAGTAGGA',\n",
    "'i5D02':'TGGCATGA',\n",
    "'i5D03':'CATGATGA',\n",
    "'i5D04':'ATCCAGGA',\n",
    "'i5D05':'CAATGACA',\n",
    "'i5D06':'CGGAATCA',\n",
    "'i5D07':'TTGGTACA',\n",
    "'i5D08':'CTACTGTA',\n",
    "'i5D09':'TCTTCTGA',\n",
    "'i5D10':'AGTTCGTA',\n",
    "'i5D11':'ACCTACGA',\n",
    "'i5D12':'TGAACGGA',\n",
    "'i5E01':'CAACACTA',\n",
    "'i5E02':'ATAGAGCA',\n",
    "'i5E03':'ACATTCCA',\n",
    "'i5E04':'TGCAACCA',\n",
    "'i5E05':'TCCTCCTA',\n",
    "'i5E06':'CTCGTCTA',\n",
    "'i5E07':'AGAGATGA',\n",
    "'i5E08':'CCTAAGTA',\n",
    "'i5E09':'CTGGAAGA',\n",
    "'i5E10':'TGCCAATA',\n",
    "'i5E11':'TCTACGCA',\n",
    "'i5E12':'ACTGTGGA',\n",
    "'i5F01':'TGATTGCA',\n",
    "'i5F02':'CTTACGGA',\n",
    "'i5F03':'CACATGCA',\n",
    "'i5F04':'TTCAGTCA',\n",
    "'i5F05':'AAGGACCA',\n",
    "'i5F06':'TGGTGGTA',\n",
    "'i5F07':'CTCAGGTA',\n",
    "'i5F08':'TCGAGACA',\n",
    "'i5F09':'CCAGAACA',\n",
    "'i5F10':'TTACACGA',\n",
    "'i5F11':'CAGACCTA',\n",
    "'i5F12':'CGCATATA',\n",
    "'i5G01':'ACTGACTA',\n",
    "'i5G02':'ACAGGTCA',\n",
    "'i5G03':'AACACGGA',\n",
    "'i5G04':'TGTCTGGA',\n",
    "'i5G05':'CTAGTTCA',\n",
    "'i5G06':'CTATCCTA',\n",
    "'i5G07':'CATCAGCA',\n",
    "'i5G08':'AGTGGATA',\n",
    "'i5G09':'TCACCATA',\n",
    "'i5G10':'CGATCTCA',\n",
    "'i5G11':'AATTGCGA',\n",
    "'i5G12':'AGTCCTCA',\n",
    "'i5H01':'TTGACCGA',\n",
    "'i5H02':'ATGTCTCA',\n",
    "'i5H03':'AACGTAGA',\n",
    "'i5H04':'CTTGCACA',\n",
    "'i5H05':'TTCATGGA',\n",
    "'i5H06':'CCATGGTA',\n",
    "'i5H07':'TTAGCGTA',\n",
    "'i5H08':'CCATTAGA',\n",
    "'i5H09':'CTAAGCGA',\n",
    "'i5H10':'TCGTAGCA',\n",
    "'i5H11':'CAAGTGGA',\n",
    "'i5H12':'TCCGTTCA'\n",
    "}\n",
    "\n",
    "i5Dict = {\n",
    "'i5A01':'GAGGTAGT',\n",
    "'i5A02':'GCTTAACT',\n",
    "'i5A03':'GCAATTCT',\n",
    "'i5A04':'TCCTCACT',\n",
    "'i5A05':'AGTTAGCT',\n",
    "'i5A06':'TCATGGCT',\n",
    "'i5A07':'AGCCTAAT',\n",
    "'i5A08':'AACACTCT',\n",
    "'i5A09':'AGAGCTCT',\n",
    "'i5A10':'GGCCATAT',\n",
    "'i5A11':'TAGAACGT',\n",
    "'i5A12':'GGTTGAAT',\n",
    "'i5B01':'GATCCTAT',\n",
    "'i5B02':'TGGTCTCT',\n",
    "'i5B03':'TCCAGTCT',\n",
    "'i5B04':'TAGCGTAT',\n",
    "'i5B05':'GCGATCAT',\n",
    "'i5B06':'AGCTACAT',\n",
    "'i5B07':'ACTCATCT',\n",
    "'i5B08':'TTGCGAGT',\n",
    "'i5B09':'AAGCGACT',\n",
    "'i5B10':'TGCACGAT',\n",
    "'i5B11':'TATGCGGT',\n",
    "'i5B12':'TCGTCGAT',\n",
    "'i5C01':'ACACTTGT',\n",
    "'i5C02':'TACGGCAT',\n",
    "'i5C03':'TCTCGTGT',\n",
    "'i5C04':'GCAGATGT',\n",
    "'i5C05':'TCCGTTGT',\n",
    "'i5C06':'TTCGCAGT',\n",
    "'i5C07':'GGAATAGT',\n",
    "'i5C08':'GCAACAAT',\n",
    "'i5C09':'ACTGCGAT',\n",
    "'i5C10':'TGAGTACT',\n",
    "'i5C11':'ACACGAAT',\n",
    "'i5C12':'GTGGCAAT',\n",
    "'i5D01':'GTCATCCT',\n",
    "'i5D02':'ACCGTACT',\n",
    "'i5D03':'GTACTACT',\n",
    "'i5D04':'TAGGTCCT',\n",
    "'i5D05':'GTTACTGT',\n",
    "'i5D06':'GCCTTAGT',\n",
    "'i5D07':'AACCATGT',\n",
    "'i5D08':'GATGACAT',\n",
    "'i5D09':'AGAAGACT',\n",
    "'i5D10':'TCAAGCAT',\n",
    "'i5D11':'TGGATGCT',\n",
    "'i5D12':'ACTTGCCT',\n",
    "'i5E01':'GTTGTGAT',\n",
    "'i5E02':'TATCTCGT',\n",
    "'i5E03':'TGTAAGGT',\n",
    "'i5E04':'ACGTTGGT',\n",
    "'i5E05':'AGGAGGAT',\n",
    "'i5E06':'GAGCAGAT',\n",
    "'i5E07':'TCTCTACT',\n",
    "'i5E08':'GGATTCAT',\n",
    "'i5E09':'GACCTTCT',\n",
    "'i5E10':'ACGGTTAT',\n",
    "'i5E11':'AGATGCGT',\n",
    "'i5E12':'TGACACCT',\n",
    "'i5F01':'ACTAACGT',\n",
    "'i5F02':'GAATGCCT',\n",
    "'i5F03':'GTGTACGT',\n",
    "'i5F04':'AAGTCAGT',\n",
    "'i5F05':'TTCCTGGT',\n",
    "'i5F06':'ACCACCAT',\n",
    "'i5F07':'GAGTCCAT',\n",
    "'i5F08':'AGCTCTGT',\n",
    "'i5F09':'GGTCTTGT',\n",
    "'i5F10':'AATGTGCT',\n",
    "'i5F11':'GTCTGGAT',\n",
    "'i5F12':'GCGTATAT',\n",
    "'i5G01':'TGACTGAT',\n",
    "'i5G02':'TGTCCAGT',\n",
    "'i5G03':'TTGTGCCT',\n",
    "'i5G04':'ACAGACCT',\n",
    "'i5G05':'GATCAAGT',\n",
    "'i5G06':'GATAGGAT',\n",
    "'i5G07':'GTAGTCGT',\n",
    "'i5G08':'TCACCTAT',\n",
    "'i5G09':'AGTGGTAT',\n",
    "'i5G10':'GCTAGAGT',\n",
    "'i5G11':'TTAACGCT',\n",
    "'i5G12':'TCAGGAGT',\n",
    "'i5H01':'AACTGGCT',\n",
    "'i5H02':'TACAGAGT',\n",
    "'i5H03':'TTGCATCT',\n",
    "'i5H04':'GAACGTGT',\n",
    "'i5H05':'AAGTACCT',\n",
    "'i5H06':'GGTACCAT',\n",
    "'i5H07':'AATCGCAT',\n",
    "'i5H08':'GGTAATCT',\n",
    "'i5H09':'GATTCGCT',\n",
    "'i5H10':'AGCATCGT',\n",
    "'i5H11':'GTTCACCT',\n",
    "'i5H12':'AGGCAAGT'\n",
    "}\n",
    "\n",
    "i7revcomp_Dict = {\n",
    "'i7A01':'GTACGTCA',\n",
    "'i7A02':'TGCAGTTA',\n",
    "'i7A03':'ACTGTGGA',\n",
    "'i7A04':'GGTTAAGA',\n",
    "'i7A05':'TCACACTA',\n",
    "'i7A06':'TAGAGGTA',\n",
    "'i7A07':'GCGACATA',\n",
    "'i7A08':'TACATGCA',\n",
    "'i7A09':'GATGATGA',\n",
    "'i7A10':'TGTGTGCA',\n",
    "'i7A11':'TCGCTACA',\n",
    "'i7A12':'AAGCTAGA',\n",
    "'i7B01':'TAGGACCA',\n",
    "'i7B02':'TCGTTGGA',\n",
    "'i7B03':'GTGTCCTA',\n",
    "'i7B04':'TCCGTATA',\n",
    "'i7B05':'TACGCTTA',\n",
    "'i7B06':'TACAACGA',\n",
    "'i7B07':'GAAGGAGA',\n",
    "'i7B08':'AAGCATCA',\n",
    "'i7B09':'TGAGATCA',\n",
    "'i7B10':'TCTAGACA',\n",
    "'i7B11':'TGCTCATA',\n",
    "'i7B12':'TAACTCCA',\n",
    "'i7C01':'GACACACA',\n",
    "'i7C02':'TTGTAGCA',\n",
    "'i7C03':'GTCTACGA',\n",
    "'i7C04':'GCATTACA',\n",
    "'i7C05':'GTTCCATA',\n",
    "'i7C06':'GAATACCA',\n",
    "'i7C07':'TCACTTGA',\n",
    "'i7C08':'ACACAAGA',\n",
    "'i7C09':'ACTACAGA',\n",
    "'i7C10':'TCTCAGGA',\n",
    "'i7C11':'ATCGTGCA',\n",
    "'i7C12':'TCCACGTA',\n",
    "'i7D01':'ATCCATGA',\n",
    "'i7D02':'TTCGCAGA',\n",
    "'i7D03':'TCGTCTCA',\n",
    "'i7D04':'ACTTACGA',\n",
    "'i7D05':'TCGGATTA',\n",
    "'i7D06':'TTGGTCTA',\n",
    "'i7D07':'ATAGCGGA',\n",
    "'i7D08':'TGTAACCA',\n",
    "'i7D09':'ACAGGCTA',\n",
    "'i7D10':'AAGTCGCA',\n",
    "'i7D11':'GAGTTCGA',\n",
    "'i7D12':'ACTCTTCA',\n",
    "'i7E01':'AAGACCTA',\n",
    "'i7E02':'ACCATCCA',\n",
    "'i7E03':'GAACGGTA',\n",
    "'i7E04':'TAGTGAGA',\n",
    "'i7E05':'GTCCAACA',\n",
    "'i7E06':'TATGGCGA',\n",
    "'i7E07':'GCTTATCA',\n",
    "'i7E08':'GTGACTGA',\n",
    "'i7E09':'ACGTGATA',\n",
    "'i7E10':'GGTTCTTA',\n",
    "'i7E11':'GATCAGCA',\n",
    "'i7E12':'GGACAATA',\n",
    "'i7F01':'AACTCTGA',\n",
    "'i7F02':'TGACCTTA',\n",
    "'i7F03':'GGAAGTGA',\n",
    "'i7F04':'AAGAAGGA',\n",
    "'i7F05':'GCAATCTA',\n",
    "'i7F06':'ACGAGTCA',\n",
    "'i7F07':'TTATGCCA',\n",
    "'i7F08':'GCCATAGA',\n",
    "'i7F09':'TAATCGGA',\n",
    "'i7F10':'TCTCGTTA',\n",
    "'i7F11':'GGTCTGTA',\n",
    "'i7F12':'ATCCGGTA',\n",
    "'i7G01':'ATGCGACA',\n",
    "'i7G02':'TGAACGCA',\n",
    "'i7G03':'TGGTATGA',\n",
    "'i7G04':'TGGACAGA',\n",
    "'i7G05':'TTGCAAGA',\n",
    "'i7G06':'ACTAGGTA',\n",
    "'i7G07':'TACCGATA',\n",
    "'i7G08':'ATTGGTGA',\n",
    "'i7G09':'TTCCTCGA',\n",
    "'i7G10':'TCATGGTA',\n",
    "'i7G11':'TATCCACA',\n",
    "'i7G12':'GACTTGTA',\n",
    "'i7H01':'ATGGAGTA',\n",
    "'i7H02':'GTCAGATA',\n",
    "'i7H03':'GAGCACTA',\n",
    "'i7H04':'GTCATTCA',\n",
    "'i7H05':'TCCTAAGA',\n",
    "'i7H06':'GATGCGTA',\n",
    "'i7H07':'TGCTTCCA',\n",
    "'i7H08':'ATGTCAGA',\n",
    "'i7H09':'GTTGCTCA',\n",
    "'i7H10':'GATATCCA',\n",
    "'i7H11':'GTTAGCGA',\n",
    "'i7H12':'GAGGTACA'\n",
    "}\n",
    "\n",
    "i7Dict = {\n",
    "'i7A01':'TGACGTAC',\n",
    "'i7A02':'TAACTGCA',\n",
    "'i7A03':'TCCACAGT',\n",
    "'i7A04':'TCTTAACC',\n",
    "'i7A05':'TAGTGTGA',\n",
    "'i7A06':'TACCTCTA',\n",
    "'i7A07':'TATGTCGC',\n",
    "'i7A08':'TGCATGTA',\n",
    "'i7A09':'TCATCATC',\n",
    "'i7A10':'TGCACACA',\n",
    "'i7A11':'TGTAGCGA',\n",
    "'i7A12':'TCTAGCTT',\n",
    "'i7B01':'TGGTCCTA',\n",
    "'i7B02':'TCCAACGA',\n",
    "'i7B03':'TAGGACAC',\n",
    "'i7B04':'TATACGGA',\n",
    "'i7B05':'TAAGCGTA',\n",
    "'i7B06':'TCGTTGTA',\n",
    "'i7B07':'TCTCCTTC',\n",
    "'i7B08':'TGATGCTT',\n",
    "'i7B09':'TGATCTCA',\n",
    "'i7B10':'TGTCTAGA',\n",
    "'i7B11':'TATGAGCA',\n",
    "'i7B12':'TGGAGTTA',\n",
    "'i7C01':'TGTGTGTC',\n",
    "'i7C02':'TGCTACAA',\n",
    "'i7C03':'TCGTAGAC',\n",
    "'i7C04':'TGTAATGC',\n",
    "'i7C05':'TATGGAAC',\n",
    "'i7C06':'TGGTATTC',\n",
    "'i7C07':'TCAAGTGA',\n",
    "'i7C08':'TCTTGTGT',\n",
    "'i7C09':'TCTGTAGT',\n",
    "'i7C10':'TCCTGAGA',\n",
    "'i7C11':'TGCACGAT',\n",
    "'i7C12':'TACGTGGA',\n",
    "'i7D01':'TCATGGAT',\n",
    "'i7D02':'TCTGCGAA',\n",
    "'i7D03':'TGAGACGA',\n",
    "'i7D04':'TCGTAAGT',\n",
    "'i7D05':'TAATCCGA',\n",
    "'i7D06':'TAGACCAA',\n",
    "'i7D07':'TCCGCTAT',\n",
    "'i7D08':'TGGTTACA',\n",
    "'i7D09':'TAGCCTGT',\n",
    "'i7D10':'TGCGACTT',\n",
    "'i7D11':'TCGAACTC',\n",
    "'i7D12':'TGAAGAGT',\n",
    "'i7E01':'TAGGTCTT',\n",
    "'i7E02':'TGGATGGT',\n",
    "'i7E03':'TACCGTTC',\n",
    "'i7E04':'TCTCACTA',\n",
    "'i7E05':'TGTTGGAC',\n",
    "'i7E06':'TCGCCATA',\n",
    "'i7E07':'TGATAAGC',\n",
    "'i7E08':'TCAGTCAC',\n",
    "'i7E09':'TATCACGT',\n",
    "'i7E10':'TAAGAACC',\n",
    "'i7E11':'TGCTGATC',\n",
    "'i7E12':'TATTGTCC',\n",
    "'i7F01':'TCAGAGTT',\n",
    "'i7F02':'TAAGGTCA',\n",
    "'i7F03':'TCACTTCC',\n",
    "'i7F04':'TCCTTCTT',\n",
    "'i7F05':'TAGATTGC',\n",
    "'i7F06':'TGACTCGT',\n",
    "'i7F07':'TGGCATAA',\n",
    "'i7F08':'TCTATGGC',\n",
    "'i7F09':'TCCGATTA',\n",
    "'i7F10':'TAACGAGA',\n",
    "'i7F11':'TACAGACC',\n",
    "'i7F12':'TACCGGAT',\n",
    "'i7G01':'TGTCGCAT',\n",
    "'i7G02':'TGCGTTCA',\n",
    "'i7G03':'TCATACCA',\n",
    "'i7G04':'TCTGTCCA',\n",
    "'i7G05':'TCTTGCAA',\n",
    "'i7G06':'TACCTAGT',\n",
    "'i7G07':'TATCGGTA',\n",
    "'i7G08':'TCACCAAT',\n",
    "'i7G09':'TCGAGGAA',\n",
    "'i7G10':'TACCATGA',\n",
    "'i7G11':'TGTGGATA',\n",
    "'i7G12':'TACAAGTC',\n",
    "'i7H01':'TACTCCAT',\n",
    "'i7H02':'TATCTGAC',\n",
    "'i7H03':'TAGTGCTC',\n",
    "'i7H04':'TGAATGAC',\n",
    "'i7H05':'TCTTAGGA',\n",
    "'i7H06':'TACGCATC',\n",
    "'i7H07':'TGGAAGCA',\n",
    "'i7H08':'TCTGACAT',\n",
    "'i7H09':'TGAGCAAC',\n",
    "'i7H10':'TGGATATC',\n",
    "'i7H11':'TCGCTAAC',\n",
    "'i7H12':'TGTACCTC'    \n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false
   },
   "source": [
    "Define barcode plateview image pairs, which log well ID relationships to barcode sequence (for dual indexed Workflows A & B). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deletable": false,
    "editable": false,
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Workflow A\n",
    "plateview_image_list_A = ['SampleSheet_img/i7_WorkflowA.png', 'SampleSheet_img/i5_WorkflowA.png']\n",
    "# Workflow B\n",
    "plateview_image_list_B = ['SampleSheet_img/i7_WorkflowB.png','SampleSheet_img/i5_WorkflowB.png']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### II. Define user-specified variables\n",
    "\n",
    "A user defines input variables by entering individual lines of text at the Jupyter interface.  \n",
    "\n",
    "Up to 6 variables will now be defined as inputs (see **'User inputs'** above). \n",
    "\n",
    "You will be prompted for the following user-specific information:\n",
    "   - Illumina® Indexed Sequencing Workflow ('A' or 'B'?)\n",
    "   - Absolute path to output directory and filename for Sample Sheet\n",
    "   - Investigator Name\n",
    "   - Project Name\n",
    "   - Single-end (SE) or Paired-end (PE) sequencing run?  How many reads?\n",
    "   - List of sample:barcode relationships\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false
   },
   "source": [
    "Specify Illumina Indexed Sequencing Workflow ('A' vs. 'B')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deletable": false,
    "editable": false,
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Specify Illumina Indexed Sequencing Workflow ('A' vs. 'B')\n",
    "workflow = input(\"\"\"\n",
    "---------------------------------------------\n",
    "Illumina® Indexed Sequencing Workflow (A or B)\n",
    "---------------------------------------------\n",
    "    \n",
    "Illumina® Indexed Sequencing for dual-indexed (Paired End) runs uses one of two different Workflows (A or B),\n",
    "defined by whether (A) both index sequences (i7 and i5) are sequenced by primers that anneal to the 'Read 1' strand,\n",
    "or (B) i7 index is sequenced by a primer that anneals to the 'Read 1' strand, whereas i5 index is sequenced by a\n",
    "primer that anneals to the 'Read 2' strand.\n",
    "\n",
    "Awareness of the different Workflows for sequencing the indices is critical, because Workflow determines whether\n",
    "the 5'->3' index sequence (i5 or i7) is returned as a 'forward'sequence that reads just like the index sequence as\n",
    "it occurs in indexing primers during library preparation, or whether the 5'->3' index sequence is returned as a\n",
    "'reverse complement' relative to the index sequence as it occurs in indexing primers during library preparation.\n",
    "Fundamentally, Workflow determines the 5'->3' nucleotide sequences entered for i7 and i5 indices in an Illumina®\n",
    "Sample Sheet, essential for faithful demultiplexing of samples.\n",
    "    \n",
    "As of June 2019, Illumina® sequencing instruments use the two dual-indexed Workflows as follows:\n",
    "      \n",
    "    Workflow A: MiSeq, NovaSeq 6000, HiSeq 2500, HiSeq 2000\n",
    "    -------------------------------------------------------\n",
    "    --> i7 index sequence is recovered as 'reverse complement' relative to i7 sequence\n",
    "        as it occurs in primers used for library construction.\n",
    "    --> i5 index sequence is recovered as it occurs in primers used for library construction.\n",
    "      \n",
    "    Workflow B: iSeq100, MiniSeq, NextSeq, HiSeqX, HiSeq 4000, HiSeq 3000\n",
    "    ---------------------------------------------------------------------\n",
    "    --> i7 index sequence is recovered as 'reverse complement' relative to i7 sequence\n",
    "        as it occurs in primers used for library construction.\n",
    "    --> i5 index sequence is recovered as 'reverse complement' relative to i5 sequence\n",
    "        as it occurs in primers used for library construction.\n",
    "      \n",
    "    Please confirm the Workflow appropriate for your sequencing application.\n",
    "    \n",
    "    Note, if your application is strictly single-indexed (Single End run using only i7 indices), choose Workflow 'A'.  \n",
    "      \n",
    "    Enter 'A' or 'B' to specify the Workflow, and therefore the index sequence orientations,\n",
    "    appropriate for your Sample Sheet:  \"\"\")\n",
    "\n",
    "if workflow in ('A', 'B'):\n",
    "    pass\n",
    "else:\n",
    "    while workflow not in ('A', 'B'):\n",
    "        workflow = input(\"\"\"\n",
    "    Type 'A' or 'B', or click on the menu 'Kernel -> Restart' to clear and reset:  \"\"\")\n",
    "\n",
    "print(r\"\"\"\n",
    "PLATEVIEWS showing relationships between integer (1-96), position in a 96-well plate (A01-H12), and 8-bp sequence\n",
    "(as to be represented in Illumina® Sample Sheet) will now be displayed.\"\"\")\n",
    "\n",
    "input(r\"\"\"\n",
    "Please press 'Enter' to continue: \"\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deletable": false,
    "editable": false,
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "print(\"\"\"\n",
    "PLATEVIEWS: correspondence between barcode well ID ('A01'-'H12') and number ('1'-'96')\n",
    "---------------------------------------------------------------------------------------------------\n",
    "SampleSheet.py requires entry of i7 and i5 identities in number format ('1'-'96') corresponding to well ID,\n",
    "making a console plateview potentially useful.\"\"\")\n",
    "\n",
    "# Display console PLATEVIEWs.\n",
    "if workflow == 'A':\n",
    "    print(\"\"\"\n",
    "WORKFLOW A.  A console view of 8-bp barcode sequences (indices) will now be displayed.\"\"\")\n",
    "    input(\"\"\"\n",
    "Press 'Enter' to display i7 PLATEVIEW\n",
    "    \"\"\")\n",
    "    print(\"\"\"PLATEVIEW:  Barcode sequences, i7  (5'->3')\n",
    "\n",
    "Please note, each 8-bp barcode sequence displayed in this image is the sequence to be used in a Workflow A\n",
    "Sample Sheet barcode field. The displayed sequence is the reverse complement of the barcode sequence as it\n",
    "occurs in the i7 primer.\n",
    "\"\"\")\n",
    "    display(Image(filename=plateview_image_list_A[0]))\n",
    "    input(\"\"\"Press 'Enter' to display i5 PLATEVIEW\"\"\")\n",
    "    print(\"\"\"\n",
    "PLATEVIEW:  Barcode sequences, i5  (5'->3')\n",
    "\n",
    "Please note, each 8-bp barcode sequence as displayed in this table is the sequence to be used in a Workflow A\n",
    "Sample Sheet barcode field. The displayed sequence is identical to the barcode sequence as it occurs in the\n",
    "i5 primer.\n",
    "\"\"\")\n",
    "    display(Image(filename=plateview_image_list_A[1]))\n",
    "elif workflow == 'B':\n",
    "    print(\"\"\"\n",
    "WORKFLOW B.  A console view of 8-bp barcode sequences (indices) will now be displayed.\"\"\")\n",
    "    input(\"\"\"\n",
    "Press 'Enter' to display i7 PLATEVIEW\n",
    "    \"\"\")\n",
    "    print(\"\"\"PLATEVIEW:  Barcode sequences, i7  (5'->3')\n",
    "\n",
    "Please note, each 8-bp barcode sequence as displayed in this table is the sequence to be used in a Workflow B\n",
    "Sample Sheet barcode field. The displayed sequence is the reverse complement of the barcode sequence as it\n",
    "occurs in the i7 primer.\n",
    "\"\"\")\n",
    "    display(Image(filename=plateview_image_list_B[0]))\n",
    "    input(\"\"\"Press 'Enter' to display i5 PLATEVIEW\"\"\")   \n",
    "    print(\"\"\"\n",
    "PLATEVIEW:  Barcode sequences, i5  (5'->3')\n",
    "\n",
    "Please note, each 8-bp barcode sequence as displayed in this table is the sequence to be used in a Workflow B\n",
    "Sample Sheet barcode field. The displayed sequence is the reverse complement of the barcode sequence as it\n",
    "occurs in the i5 primer.\n",
    "\"\"\")\n",
    "    display(Image(filename=plateview_image_list_B[1]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deletable": false,
    "editable": false,
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Specify user inputs:\n",
    "# (1) Indicate where the output Sample Sheet file should go (future .csv filename and absolute path).\n",
    "print(r\"\"\"  \n",
    "---------------------------------------------------------------------------\n",
    "Sample Sheet file name and location (absolute path to future .csv filename)\n",
    "---------------------------------------------------------------------------\n",
    "\n",
    "Enter the name of the .csv file you'd like to create as your Sample Sheet, with an absolute path to its location.\n",
    "\n",
    "The .csv file should not exist yet -- it will be created as an output of this script.\n",
    " \n",
    "To create the .csv file in a directory where you can find it, enter an *absolute path* to where you would like\n",
    "this file to be created, using only forward slashes ('/') to indicate directory separations.\n",
    "\n",
    "Example: if your target file name is 'SampleSheet.csv', and you'd like to create that file in a directory\n",
    "that is accessed with an absolute path of '/Users/myname/Illumina/SampleSheet.csv' (Mac) or\n",
    "'C:\\Users\\myname\\Illumina\\SampleSheet.csv' (PC), enter '/Users/myname/Illumina/SampleSheet.csv' (Mac) or\n",
    "'C:/Users/myname/Illumina/SampleSheet.csv' (PC) at the command line prompt. Replace 'myname, etc.' with the\n",
    "appropriate intervening directory identifiers. Do *not* flank your entry with quotation marks (') at the command-line. \n",
    "    \n",
    "Alternatively, simply enter a target file name (e.g., 'SampleSheet.csv') and run this script from\n",
    "within a directory where you'd like to output this file.\n",
    "\"\"\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deletable": false,
    "editable": false,
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "filename = input(r\"\"\"    -----> File name and path: \"\"\")\n",
    "\n",
    "# Wait to actually create the file until later in the script, in case there is a need to restart the script for a given file.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deletable": false,
    "editable": false,
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# [Header] details: specify InvestigatorName & ProjectName\n",
    "print(r\"\"\"  \n",
    "-----------------------------------------------------------------\n",
    "Sample Sheet inputs: [Header], [Reads], and [Data] specifications\n",
    "-----------------------------------------------------------------\n",
    "      \n",
    "........................................................\n",
    "[Header] details: specify InvestigatorName & ProjectName\n",
    "    \n",
    "To specify InvestigatorName & ProjectName, enter text for each directly at the command line,\n",
    "separated by a comma ('InvestigatorName, ProjectName').\n",
    "    \n",
    "When both text entries are entered, press ‘enter’ again to proceed in the script.\n",
    "To skip text entries for these fields, simply press ‘enter’ until the next prompt appears (if you skip\n",
    "entry now, 'NA' will be entered in these fields in the output file).\n",
    "\n",
    "Example: if your InvestigatorName is 'Dorothy Gale' and ProjectName is 'Sequences', enter\n",
    "'Dorothy Gale, Sequences'.\n",
    "\"\"\")\n",
    "header = input(r\"\"\"    -----> [Header] details. InvestigatorName & ProjectName: \"\"\")\n",
    "\n",
    "if type(header.split(',')) is list:\n",
    "    InvestigatorName = header.split(',')[0].strip()\n",
    "    ProjectName = header.split(',')[1].strip()\n",
    "else:\n",
    "    InvestigatorName = 'NA'\n",
    "    ProjectName = 'NA'\n",
    "    \n",
    "# [Reads] details: specify Single-End vs. Paired-End\n",
    "print(r\"\"\"\n",
    ".................................................................................................\n",
    "[Reads] details: specify whether sequencing is Single-End or Paired-End, and the number of cycles\n",
    "    \n",
    "To specify Single-End vs. Paired-End format and number of cycles, enter text directly at the command line,\n",
    "on a single line. Indicate Single-End (SE) or Paired-End (PE), followed by the number of cycles for each read.\n",
    "Separate values by comma(s).\n",
    "\n",
    "When text is entered, press ‘enter’ again to proceed in the script.\n",
    "\n",
    "Examples:\n",
    "If you are performing a Paired-End run with 151 cycles in reads 1 & 2, enter\n",
    "'PE, 151, 151'.\n",
    "\n",
    "If you are performing a Single-End run with 151 cycles in read 1, enter\n",
    "'SE, 151'.\n",
    "\"\"\")\n",
    "reads = input(r\"\"\"    ----> [Reads] details: \"\"\")\n",
    "\n",
    "reads_verification = '0'\n",
    "while reads_verification == '0':\n",
    "    if reads.split(',')[0].strip() in ('SE', 'PE'):\n",
    "        readslist = [i.strip() for i in reads.split(',')]\n",
    "        if readslist[0] == 'SE':\n",
    "            if len(readslist) == 2:\n",
    "                readsvalue = readslist[1] \n",
    "                readstype = readslist[0]\n",
    "                reads_verification = '1'\n",
    "            else:\n",
    "                reads = input(\"\"\"\n",
    "    You indicated 'SE' run, but indicated an incommensurate value for # of reads (should be exactly one cycle #\n",
    "    value); please correct your entry. Indicate only cycle # for insert read(s) (do not include index read cycles).\n",
    "    Type 'PE' or 'SE' followed by appropriate cycle number(s), or click on the menu 'Kernel -> Restart' to clear\n",
    "    and reset:  \"\"\")\n",
    "        elif readslist[0] == 'PE':\n",
    "            if len(readslist) == 3:\n",
    "                readsvalue = readslist[1]+'\\n'+readslist[2]\n",
    "                readstype = readslist[0]\n",
    "                reads_verification = '1'\n",
    "            else:\n",
    "                reads = input(\"\"\"\n",
    "    You indicated 'PE' run, but indicated an incommensurate value for # of reads (should be exactly two cycle #\n",
    "    values); please correct your entry. Indicate only cycle # for insert read(s) (do not include index read cycles).\n",
    "    Type 'PE' or 'SE' followed by appropriate cycle number(s), or click on the menu 'Kernel -> Restart' to clear\n",
    "    and reset:  \"\"\")\n",
    "    else:\n",
    "        while reads.split(',')[0].strip() not in ('SE', 'PE'):\n",
    "            reads = input(\"\"\"\n",
    "    Type 'PE' or 'SE' followed by appropriate cycle numbers, or click on the menu 'Kernel -> Restart' to clear\n",
    "    and reset:  \"\"\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deletable": false,
    "editable": false,
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Check for compatibility between SE/PE and Workflow A/B\n",
    "if readstype == 'PE':\n",
    "    pass\n",
    "elif readstype == 'SE':\n",
    "    if workflow == 'A':\n",
    "        pass\n",
    "    elif workflow == 'B':\n",
    "        compatibility = input(\"\"\"\n",
    "    ***** CAUTION: ***** \n",
    "    ***** Dual-indexed runs (using both i7 and i5 barcodes) may use Workflow A or B;\n",
    "    single-indexed runs (using only i7 barcodes) must be specified as Workflow A. *****\n",
    "    \n",
    "    You selected Workflow B, a dual-indexed (Paired-End) sequencing format, while also indicating that you are\n",
    "    setting up a Sample Sheet for a single-indexed (Single-End) run. Illumina Single-End runs use exclusively the i7 index\n",
    "    with Workflow A (i7 index is read on the same molecule as Read 1). \n",
    "    \n",
    "    Please quit this script session by :  \"\"\")\n",
    "        if compatibility == 'Exit':\n",
    "            exit(0)\n",
    "        else:\n",
    "            input(\"\"\"\n",
    "    ***** CAUTION: *****\n",
    "    Workflow 'B' and 'SE' sequencing specifications are not compatible.\n",
    "    You may now proceed temporarily in the script, but will encounter an error upon trying to populate\n",
    "    non-existent i5 barcodes into Sample Sheet output file. Press Enter to continue...\n",
    "    \"\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# [Data] details: specify list of plate names, i7 barcode range, and i5 barcode used for each plate.\n",
    "if readstype == 'PE':\n",
    "    print(r\"\"\"\n",
    "........................................................................\n",
    "[Data] details: specify relationships between sample names and barcodes. \n",
    "\n",
    "You will now be asked to enter a text-only table of plate names with their corresponding\n",
    "i7 barcode range and i5 barcode, for dual-indexed (Paired-End) sequencing. \n",
    " \n",
    "Specifics:\n",
    "    * Each line corresponds to samples barcoded across a single 96-well plate.\n",
    "    * Each line is expanded in the output Sample Sheet based on the 'plate name' and specified number of\n",
    "        barcoded sample wells; individual wells are assigned unique sample IDs in the Sample Sheet based on their\n",
    "        unique combination of 'plate name' and 'i7 barcode' well ID.\n",
    "        * 'Plate name' identifies a single 96-well plate identifier and must be unique.\n",
    "            * Any letter, digit, and punctuation characters are acceptable in names,\n",
    "              excluding underscores ('_') and semicolons (';'), which must *not* be used. \n",
    "        * 'i7 barcode' identifies an individual well\n",
    "          (entry will be a numeric range, any # range up to '1-96')\n",
    "        * 'i5 barcode' identifies all wells in a single plate\n",
    "          (entry will be a single number, any # in '1' to '96'.)\n",
    "        * For i7 and i5 barcode identifiers, use only the *range* (e.g., '1-96') or *number* (e.g., '1'-'96')\n",
    "          that corresponds to a given index. Refer to i5 and i7 96-well plate sequences\n",
    "          (displayed earlier as console PLATEVIEWS), if needed.\n",
    "        * Fields (plate name, i7 index range, i5 index) are comma-separated.\n",
    "        * You may manually enter or paste up to 96 lines that specify sample name-barcode relationships.\n",
    "        * End each plate description line with a semi-colon (';'), except for the final line.\n",
    "\n",
    "Example: imagine that you have 310 samples barcoded across four 96-well plates (some plates containing 96 samples,\n",
    "some plates containing fewer than 96 samples). Plate 1 used unique i7 barcodes '1-96' ('i7A01-i7H12') across 96\n",
    "samples + i5 barcode '4' ('i5A05') for all 96 samples; Plate 2 used the same i7 barcode range + i5 barcode '5'\n",
    "('i5A05') for its 96 samples, Plate 3 used unique barcodes '1-50' ('i7A01-i7E02') across 50 samples + i5 barcode '2'\n",
    "('i5A02') for all 50 samples; Plate 4 used unique barcodes '1-68' ('i7A01-i7F08') across 68 samples + i5 barcode '34'\n",
    "('i5C10') for all 68 samples. You would enter text, line by line at the command line, that resembles this:\n",
    "\n",
    "        DG-1, 1-96, 1;\n",
    "        DG-2, 1-96, 9;\n",
    "        DG-3, 1-50, 78;\n",
    "        DG-4, 1-68, 34\n",
    "\n",
    "When you're done entering plates and their indices, press 'Enter' again to proceed in the script. \n",
    "\"\"\")\n",
    "elif readstype == 'SE':\n",
    "    print(r\"\"\"\n",
    ".......................................................................\n",
    "[Data] details: specify relationships between sample names and barcode. \n",
    "\n",
    "You will now be asked to enter a text-only table of plate name(s) with their corresponding\n",
    "i7 barcode distribution across wells, for single-indexed (Single-End) sequencing.\n",
    " \n",
    "Specifics:\n",
    "    * Each line corresponds to samples barcoded across a single 96-well plate.\n",
    "    * Each line is expanded in the output Sample Sheet based on the 'plate name' and specified number of\n",
    "        barcoded sample wells; individual wells are assigned unique sample IDs in the Sample Sheet based on their\n",
    "        unique combination of 'plate name' and 'i7 barcode' well ID.\n",
    "        * 'Plate name' identifies a single 96-well plate identifier and must be unique.\n",
    "            * Any letter, digit, and punctuation characters are acceptable in names,\n",
    "              excluding underscores ('_') and semicolons (';'), which must *not* be used. \n",
    "        * 'i7 barcode' identifies an individual well\n",
    "          (entry will be a numeric range, any # range up to '1-96')\n",
    "        * For i7 barcode identifiers, use only the *range* (e.g., '1-96') or *number* (e.g., '1'-'96')\n",
    "          that corresponds to a given index. Refer to i7 96-well plate sequences\n",
    "          (displayed earlier as console PLATEVIEWS), if needed.\n",
    "        * Fields (plate name, i7 index range) are comma-separated.\n",
    "        * You may manually enter or paste up to 96 lines that specify sample name-barcode relationships.\n",
    "        * End each plate description line with a semi-colon (';'), except for the final line.\n",
    "\n",
    "Example: imagine that you have 310 samples barcoded across four 96-well plates (some plates containing 96 samples,\n",
    "some plates containing fewer than 96 samples). Each plate contains a different type of amplicon, but i7 indices\n",
    "are re-used for different amplicons across plates. Plate 1 used unique i7 barcodes '1-96' ('i7A01-i7H12') across\n",
    "96 samples; Plate 2 used the same i7 barcode range for its 96 samples; Plate 3 used unique barcodes '1-50'\n",
    "('i7A01-i7E02') across 50 samples; Plate 4 used unique barcodes '1-68' ('i7A01-i7F08') across 68 samples.\n",
    "You would enter text, line by line at the command line, that resembles this:\n",
    "\n",
    "        DG-1, 1-96;\n",
    "        DG-2, 1-96;\n",
    "        DG-3, 1-50;\n",
    "        DG-4, 1-68\n",
    "\n",
    "When you're done entering plates and their indices, press 'enter' again to proceed in the script.\n",
    "\"\"\")\n",
    "\n",
    "data_input = input(r'    -----> [Data] details: ')    \n",
    "\n",
    "input_list = [i.strip(';').strip() for i in data_input.split(';')]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deletable": false,
    "editable": false,
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Double-check whether entries look good:\n",
    "print(\"\"\"\n",
    "---------------------------------------------------------------\n",
    "Preparation for output:\n",
    "Please double-check that your inputs were recorded as expected.\n",
    "---------------------------------------------------------------\"\"\")\n",
    "print(\"\"\"\n",
    "Your Workflow was recorded as:\n",
    "\"\"\")\n",
    "print(workflow)\n",
    "print(\"\"\"\n",
    "Your filepath and name were recorded as:\n",
    "\"\"\")\n",
    "print(filename)\n",
    "print(\"\"\"\n",
    "Your [Header] InvestigatorName and ProjectName were recorded as:\n",
    "\"\"\")\n",
    "print(\"InvestigatorName,\"+InvestigatorName+\"\\n\"+\"ProjectName,\"+ProjectName)\n",
    "print(\"\"\"\n",
    "Your [Reads] were recorded as:\n",
    "\"\"\")\n",
    "print(\"[Reads]\\n\"+readsvalue)\n",
    "print(\"\"\"\n",
    "Your [Data] input list was recorded as:\n",
    "\"\"\") \n",
    "for input_str in input_list:\n",
    "    print(input_str)\n",
    "\n",
    "check = input(\"\"\"\n",
    "Is this list accurately recorded? Type 'Y' or 'N': \n",
    "\"\"\")\n",
    "    \n",
    "if check == 'Y':\n",
    "    pass\n",
    "elif check == 'N':\n",
    "    checkup = input(\"\"\"\n",
    "If you have corrections to make, click on the menu 'Kernel -> Restart' to quit the active session and begin again.\n",
    "Otherwise, press 'Enter' to continue in the script.\n",
    "\"\"\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false
   },
   "source": [
    "### III. Expand \\[Data\\] entry to establish sample-barcode relationships   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "deletable": false,
    "editable": false,
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Log total user interaction time duration \n",
    "interactionDuration = str(datetime.now() - initialTime).split(':')[0]+' hr|'+str(datetime.now() - initialTime).split(':')[1]+' min|'+str(datetime.now() - initialTime).split(':')[2].split('.')[0]+' sec|'+str(datetime.now() - initialTime).split(':')[2].split('.')[1]+' microsec'\n",
    "\n",
    "# Begin script operation time clock\n",
    "startTime = datetime.now()\n",
    "\n",
    "# Construct [Data] Section of Sample Sheet:\n",
    "# get plate names\n",
    "plate_names = [i.partition(',')[0].strip() for i in input_list]\n",
    "\n",
    "# get i7 indices\n",
    "if readstype == 'PE':\n",
    "    i7_indexrange = [i.partition(',')[-1].rpartition(',')[0].strip() for i in input_list]\n",
    "elif readstype == 'SE':\n",
    "    i7_indexrange = [i.partition(',')[2].strip() for i in input_list]\n",
    "    \n",
    "if readstype == 'PE':\n",
    "# get i5 indices\n",
    "    i5_indexrange = [i.rpartition(',')[-1].strip() for i in input_list]\n",
    "    \n",
    "# expand i7 index list.  Generates a list of lists containing tuples.\n",
    "i7_index_expansion = []\n",
    "for i in i7_indexrange:\n",
    "    i7_index_expansion.append(i7_well_IDs[int(i.partition('-')[0].strip())-1:int(i.rpartition('-')[-1].strip())])\n",
    "    \n",
    "if readstype == 'PE':\n",
    "# expand i5 index list.  Generates a list of lists containing tuples.\n",
    "    i5_index_expansion = []\n",
    "    for i in i5_indexrange:\n",
    "        i5_index_expansion.append(i5_well_IDs[int(i.partition('-')[0].strip())-1:int(i.rpartition('-')[-1].strip())])\n",
    "        \n",
    "# give each plate its own list of expanded i7 and i5 IDs\n",
    "expanded = []\n",
    "if readstype == 'PE':\n",
    "    for i in range(0, len(plate_names)):\n",
    "        expanded.append([plate_names[i], sorted(i[1] for i in i7_index_expansion[i]), sorted(i[1] for i in i5_index_expansion[i])])\n",
    "elif readstype == 'SE':\n",
    "    for i in range(0, len(plate_names)):\n",
    "        expanded.append([plate_names[i], sorted(i[1] for i in i7_index_expansion[i])])\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": false,
    "editable": false
   },
   "source": [
    "### IV. Create Sample Sheet and populate with content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "# Create file object (f) in the target directory, with the filename initially entered at the start of the script:\n",
    "filepath = Path(filename)\n",
    "f = open(filepath, 'a')\n",
    "f.close()\n",
    "\n",
    "# Use print redirection to write to target file, in append mode (prepare entire Sample Sheet)\n",
    "# Write to [Header], [Reads] sections\n",
    "with open(filepath, 'a') as f:\n",
    "    if readstype == 'PE':\n",
    "        print(\"\"\"[Header]\n",
    "IEMFileVersion,4\\n\"\"\" +\n",
    "\"InvestigatorName,\" + InvestigatorName +\n",
    "\"\\nProjectName,\" + ProjectName +\n",
    "\"\\nDate,\" + (time.strftime(\"%m/%d/%Y\")) + \n",
    "\"\"\"\\nWorkflow,GenerateFASTQ\n",
    "Application,FASTQ Only\n",
    "Assay,Nextera\"\"\" +\n",
    "\"\\nDescription,Sequencing Workflow \" + workflow +\n",
    "\"\"\"\\nChemistry,Amplicon\n",
    "\n",
    "[Reads]\\n\"\"\" +\n",
    "readsvalue +\n",
    "\"\"\"\\n\\n[Settings]\n",
    "ReverseComplement,0\n",
    "Adapter,CTGTCTCTTATACACATCT\n",
    "\n",
    "[Data]\n",
    "Sample_ID,Sample_Name,I7_Index_ID,index,I5_Index_ID,index2\"\"\", file = f)\n",
    "    elif readstype == 'SE':\n",
    "         print(\"\"\"[Header]\n",
    "IEMFileVersion,4\\n\"\"\" +\n",
    "\"InvestigatorName,\" + InvestigatorName +\n",
    "\"\\nProjectName,\" + ProjectName +\n",
    "\"\\nDate,\" + (time.strftime(\"%m/%d/%Y\")) + \n",
    "\"\"\"\\nWorkflow,GenerateFASTQ\n",
    "Application,FASTQ Only\n",
    "Assay,Nextera\n",
    "Description,Sequencing\n",
    "Chemistry,Amplicon\n",
    "\n",
    "[Reads]\\n\"\"\" +\n",
    "readsvalue +\n",
    "\"\"\"\\n\\n[Settings]\n",
    "ReverseComplement,0\n",
    "Adapter,CTGTCTCTTATACACATCT\n",
    "\n",
    "[Data]\n",
    "Sample_ID,Sample_Name,I7_Index_ID,index\"\"\", file = f)       \n",
    "        \n",
    "f.close()\n",
    "\n",
    "# Write to [Data] section\n",
    "if workflow == 'A':\n",
    "    if readstype == 'PE':\n",
    "        with open(filename, 'a') as f:\n",
    "            count = 1\n",
    "            for i in expanded:\n",
    "                w = 0\n",
    "                while w < len(i[1]):\n",
    "                    print(str(count) + \",\" + i[0] + \"-\" + i[1][w].split('7',1)[1] + \",\" + i[1][w] + \",\" + i7revcomp_Dict.get(i[1][w]) + \",\" + i[2][0] + \",\" + i5Dict.get(i[2][0]), file = f)\n",
    "                    w = w + 1\n",
    "                    count = count + 1\n",
    "    elif readstype == 'SE':\n",
    "        with open(filename, 'a') as f:\n",
    "            count = 1\n",
    "            for i in expanded:\n",
    "                w = 0\n",
    "                while w < len(i[1]):\n",
    "                    print(str(count) + \",\" + i[0] + \"-\" + i[1][w].split('7',1)[1] + \",\" + i[1][w] + \",\" + i7revcomp_Dict.get(i[1][w]), file = f)\n",
    "                    w = w + 1\n",
    "                    count = count + 1\n",
    "elif workflow == 'B':\n",
    "    with open(filename, 'a') as f:\n",
    "        count = 1\n",
    "        for i in expanded:\n",
    "            w = 0\n",
    "            while w < len(i[1]):\n",
    "                print(str(count) + \",\" + i[0] + \"-\" + i[1][w].split('7',1)[1] + \",\" + i[1][w] + \",\" + i7revcomp_Dict.get(i[1][w]) + \",\" + i[2][0] + \",\" + i5revcomp_Dict.get(i[2][0]), file = f)\n",
    "                w = w + 1\n",
    "                count = count + 1                  \n",
    "\n",
    "f.close()\n",
    "\n",
    "# Log script processing time duration \n",
    "processingDuration = str(datetime.now()- startTime).split(':')[0]+' hr|'+str(datetime.now() - startTime).split(':')[1]+' min|'+str(datetime.now() - startTime).split(':')[2].split('.')[0]+' sec|'+str(datetime.now() - startTime).split(':')[2].split('.')[1]+' microsec'\n",
    "\n",
    "# End of script operations\n",
    "print('\\nUser input time: '+interactionDuration)\n",
    "print('\\nSample Sheet processing time: '+processingDuration)\n",
    "print(\"\"\"\n",
    "---------------------------------------------------------------------------------------------------\n",
    "Your Sample Sheet is complete.\n",
    "The file can be found at \"\"\" + filename + \"\"\" \n",
    "\n",
    "Please verify that the Sample Sheet contents describe your intended barcode assignments to samples.\n",
    "---------------------------------------------------------------------------------------------------\n",
    "\n",
    "*end of script*\n",
    "\n",
    "\"\"\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "celltoolbar": "Edit Metadata",
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
